core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -211,7 +201,11 @@ void resetRecovery(TimeValue timeout) throws IOException, TimeoutException {
                 assert stage != RecoveryState.Stage.DONE : "recovery should not have completed when it's being reset";
                 throw new IllegalStateException("cannot reset recovery as previous attempt made it past finalization step");
             }
-            indexShard.performRecoveryRestart();
+            try {
+                indexShard.performRecoveryRestart();
+            } catch (IOException ioe) {
```

Comment:
I believe you can run this via `CancellableThreads.executeIO`, which means we won't this exception juggling